src/ir/module-utils.cpp

```diff
@@ -684,95 +684,81 @@ std::vector<HeapType> getPrivateHeapTypes(Module& wasm) {
+      count += counts.at(type);
     }
-  };
+    groupCounts.push_back(count);
+  }
 
-  // Collect the information that will be used to sort the recursion groups.
-  GroupInfoMap groupInfos;
-  for (auto& [type, _] : counts) {
-    RecGroup group = type.getRecGroup();
-    // Try to initialize a new info or get the existing info.
-    auto& info = groupInfos.insert({group, {groupInfos.size()}}).first->second;
-    // Update the reference count.
-    info.useCount += counts.at(type);
-    // Collect predecessor groups.
-    for (auto child : type.getReferencedHeapTypes()) {
-      if (!child.isBasic()) {
-        RecGroup otherGroup = child.getRecGroup();
-        if (otherGroup != group) {
-          info.preds.insert(otherGroup);
+  // Collect the reverse dependencies of each group.
+  std::vector<std::unordered_set<size_t>> depSets(groups.size());
+  for (size_t i = 0; i < groups.size(); ++i) {
+    for (auto type : groups[i]) {
+      for (auto child : type.getReferencedHeapTypes()) {
+        if (child.isBasic()) {
+          continue;
         }
+        auto childGroup = child.getRecGroup();
+        if (childGroup == groups[i]) {
+          continue;
+        }
+        depSets[groupIndices.at(childGroup)].insert(i);
       }
     }
   }
-
-  // Fix up the use counts to be averages to ensure groups are used comensurate
-  // with the amount of index space they occupy. Skip this for nominal types
-  // since their internal group size is always 1.
-  for (auto& [group, info] : groupInfos) {
-    info.useCount /= group.size();
+  TopologicalSort::Graph deps;
+  deps.reserve(groups.size());
+  for (size_t i = 0; i < groups.size(); ++i) {
+    deps.emplace_back(depSets[i].begin(), depSets[i].end());
   }
 
-  // Sort the predecessors so the most used will be visited first.
-  for (auto& [group, info] : groupInfos) {
-    info.sortedPreds.insert(
-      info.sortedPreds.end(), info.preds.begin(), info.preds.end());
-    groupInfos.sort(info.sortedPreds);
-    info.preds.clear();
-  }
+  // Experimentally determined to be pretty good for a variety of programs in
+  // different languages.
+  constexpr float childFactor = 0.25;
 
-  struct RecGroupSort : TopologicalSort<RecGroup, RecGroupSort> {
-    GroupInfoMap& groupInfos;
-    RecGroupSort(GroupInfoMap& groupInfos) : groupInfos(groupInfos) {
-      // Sort all the groups so the topological sort visits the most used first.
-      std::vector<RecGroup> sortedGroups;
-      sortedGroups.reserve(groupInfos.size());
-      for (auto& [group, _] : groupInfos) {
-        sortedGroups.push_back(group);
-      }
-      groupInfos.sort(sortedGroups);
-      for (auto group : sortedGroups) {
-        push(group);
-      }
+  // Each rec group's weight, adjusted for its size and incorporating the weight
+  // of its users.
+  std::vector<float> weights(groups.size());
+  for (size_t i = 0; i < groups.size(); ++i) {
+    weights[i] = (float)groupCounts[i] / groups[i].size();
+  }
```

Comment:
Maybe another option than dividing by group size is to compute the benefit (which is nonlinear) as we go. Example to show what I mean:
```
group A {
  a // weight 101
}
group B {
  b1 // weight 190
  b2 // weight  10
}
```
The average of group B is 100, but if the first entry gets an index smaller than the two after it, then `b1,b2,a` will be smaller than `a,b1,b2`.

But maybe this is not worth the complexity.

Reply:
A tricky thing about this is that the rate of change of the benefit changes depending on the final indices the group will occupy, but maybe an approximation would be enough here.

I tried the related technique of approximating total cost of an order using the 1/128 smearing you used in ReorderGlobals, but it turned out to produce worse results than doing a more accurate count of the costs by calculating the exact byte size of each index.

Reply:
Sounds good. Yeah, this is probably good enough as is.